### ISSUE TITLE
`Never` should not be allowed as a type argument for a type parameter with constraints

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#6557]
This is a cross-post of https://github.com/python/typing/discussions/1515, not 100% if it is a bug, but it is strange behavior.

I have a class `Sliding`, which, depending on the mode returns different things when iterating over it. As a fallback, if the mode is not known statically, I want `Sliding.__iter__` to be considered `Iterator[Any]`.

The following code gives type-errors for the unknown case, as expected, since `pyright` matches the first overload. When I add the extra overload `def __iter__(self: "Sliding[T, Never]") -> Iterator[Any]: ...` as an early fallback to catch the `Any` case, then suddenly `pyright` fails for the mode-case which is listed first in the type-var definition of `Modes` (and only for this case).

<details> <summary> full example </summary>

```python
#!/usr/bin/env python

from typing import (
    TYPE_CHECKING,
    Any,
    Generic,
    Iterable,
    Iterator,
    Literal,
    Never,
    TypeAlias,
    TypeVar,
    assert_type,
    overload,
)

T = TypeVar("T")
S: TypeAlias = Literal["slices"]
B: TypeAlias = Literal["bounds"]
M: TypeAlias = Literal["masks"]
W: TypeAlias = Literal["windows"]
Mode = TypeVar("Mode", S, B, M, W)


class Sliding(Generic[T, Mode]):
    data: list[T]
    size: int
    mode: Mode

    @overload
    def __init__(self: "Sliding[T, S]", data: list[T], size: int, mode: S) -> None: ...
    @overload
    def __init__(self: "Sliding[T, B]", data: list[T], size: int, mode: B) -> None: ...
    @overload
    def __init__(self: "Sliding[T, M]", data: list[T], size: int, mode: M) -> None: ...
    @overload
    def __init__(self: "Sliding[T, W]", data: list[T], size: int, mode: W) -> None: ...
    @overload  # fallback
    def __init__(
        self: "Sliding[T, Any]", data: list[T], size: int, mode: str
    ) -> None: ...
    def __init__(self, data, size, mode):
        self.data = data
        self.size = size
        self.mode = mode

    @overload  # early fallback
    def __iter__(self: "Sliding[T, Never]") -> Iterator[Any]: ...
    @overload
    def __iter__(self: "Sliding[T, B]") -> Iterator[tuple[T, T]]: ...
    @overload
    def __iter__(self: "Sliding[T, W]") -> Iterator[list[T]]: ...
    @overload
    def __iter__(self: "Sliding[T, S]") -> Iterator[slice]: ...
    @overload
    def __iter__(self: "Sliding[T, M]") -> Iterator[list[bool]]: ...
    @overload  # late fallback
    def __iter__(self: "Sliding[T, Any]") -> Iterator[Any]: ...
    def __iter__(self):  # pyright: ignore[reportGeneralTypeIssues]
        num = len(self.data) - self.size + 1

        match self.mode:
            case "windows":
                for i in range(num):
                    yield self.data[i : i + self.size]
            case "bounds":
                for i in range(num):
                    yield i, i + self.size
            case "slices":
                for i in range(num):
                    yield slice(i, i + self.size)
            case "masks":
                for i in range(num):
                    yield [i <= j < i + self.size for j in range(len(self.data))]
            case _:
                raise TypeError(f"Unknown mode: {self.mode}")


def test_slice() -> None:
    sampler = Sliding([1, 2, 3, 4, 5], 3, mode="slices")

    if TYPE_CHECKING:
        assert_type(sampler, Sliding[int, S])
        assert_type(sampler.__iter__(), Iterator[slice])
        assert_type(iter(sampler), Iterator[slice])
        assert_type(list(sampler), list[slice])
        s1: Iterable[slice] = sampler

    assert list(sampler) == [slice(0, 3), slice(1, 4), slice(2, 5)]


def test_window() -> None:
    sampler = Sliding([1, 2, 3, 4, 5], 3, mode="windows")

    if TYPE_CHECKING:
        assert_type(sampler, Sliding[int, W])
        assert_type(sampler.__iter__(), Iterator[list[int]])
        assert_type(iter(sampler), Iterator[list[int]])
        assert_type(list(sampler), list[list[int]])
        w1: Iterable[list[int]] = sampler

    assert list(sampler) == [[1, 2, 3], [2, 3, 4], [3, 4, 5]]


def test_bounds() -> None:
    sampler = Sliding([1, 2, 3, 4, 5], 3, mode="bounds")

    if TYPE_CHECKING:
        assert_type(sampler, Sliding[int, B])
        assert_type(sampler.__iter__(), Iterator[tuple[int, int]])
        assert_type(iter(sampler), Iterator[tuple[int, int]])
        assert_type(list(sampler), list[tuple[int, int]])
        b1: Iterable[tuple[int, int]] = sampler

    assert list(sampler) == [(0, 3), (1, 4), (2, 5)]


def test_masks() -> None:
    sampler = Sliding([1, 2, 3, 4, 5], 3, mode="masks")

    if TYPE_CHECKING:
        assert_type(sampler, Sliding[int, M])
        assert_type(sampler.__iter__(), Iterator[list[bool]])
        assert_type(iter(sampler), Iterator[list[bool]])
        assert_type(list(sampler), list[list[bool]])
        m1: Iterable[list[bool]] = sampler

    assert list(sampler) == [
        [True, True, True, False, False],
        [False, True, True, True, False],
        [False, False, True, True, True],
    ]


def test_unknown() -> None:
    mode: str = "..."
    sampler = Sliding([1, 2, 3, 4, 5], 3, mode=mode)

    if TYPE_CHECKING:
        assert_type(sampler, Sliding[int, Any])
        assert_type(sampler.__iter__(), Iterator[Any])
        assert_type(iter(sampler), Iterator[Any])
        assert_type(list(sampler), list[Any])
        u1: Iterable[Any] = sampler
```

</details>

**Comment by erictraut** [comment]
You're attempting to combine some of the uglier and less-well-specified aspects of the type system, so it's not surprising that you're struggling to get consistent results between type checkers. You're using a combination of 1) constrained TypeVars (with literals, no less!), 2) explicitly-annotated `self` types (with type variables, no less!), and 3) complex overloads. These are three features I recommend avoiding if possible, and you're using all three of them in a single code sample.

I don't fully understand what you're trying to do here, but it looks very hacky. If you're able to redesign the code, you might consider alternative approaches. For example, would it make sense to define different classes here rather than trying to make a single class that has distinct modes?

As for the behavior you're seeing in pyright, I agree there's a bug here. `Sliding[T, Never]` is an invalid type annotation and should generate an error because `Never` isn't compatible with the second (constrained) type parameter. I'll work on a fix for this.

The overload matching logic in both pyright and mypy have special handling for `Any`. In particular, if `Any` appears in an argument passed to a call and this results in an ambiguous overload selection, both type checkers will evaluate the return type of the call as `Any`. However, in your case, this isn't applicable because the `Any` appears only in the `self` argument, and this argument is eliminated from the overload signatures during binding (i.e. during the evaluation of the `sampler.__iter__` expression, before the call expression is evaluated). By the time the overload matching logic is invoked, there is no `Any` argument present. The type checker is simply choosing from among the following overloads:

```python
    @overload
    def __iter__() -> Iterator[tuple[T, T]]: ...
    @overload
    def __iter__() -> Iterator[list[T]]: ...
    @overload
    def __iter__() -> Iterator[slice]: ...
```

This is why both mypy and pyright choose the first overload in this case.

**Comment by erictraut** [comment]
> ... should be in a certain sense isomorphic

They're isomorphic in the same way that a hammer and a crescent wrench are isomorphic, but if you want to pound a nail, a hammer is definitely a more effective tool for the job ;).

If you find that a class is difficult to annotate, that's often a sign that you should go back and revisit the design. There may be a better tool for the job at hand.

I've fixed the false negative with the `Never` type argument. This will be fixed in the next release. As for the other issues you noted in this bug report, I think pyright is working as intended.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.338, which I just published. It will also be included in a future release of pylance.
